### PR TITLE
fix: SuperTokens list ignoring base_path in config

### LINKED ISSUE #408: SuperTokens list command breaks when base_path is set
## 🐛 Bug Report

When base_path is set in config.yaml, the supertokens list command fails with a 404 Not Found error:
```html
io.supertokens.cli.httpRequest.HTTPResponseException: Http error. Status Code: 404. Message: <!doctype html><html lang="en"><head><title>HTTP Status 404 â?? Not Found</title><style type="text/css">h1 {font-family:Tahoma,Arial,sans-serif;color:white;background-color:#525D76;font-size:22px;} h2 {font-family:Tahoma,Arial,sans-serif;color:white;background-color:#525D76;font-size:16px;} h3 {font-family:Tahoma,Arial,sans-serif;color:white;background-color:#525D76;font-size:14px;} body {font-family:Tahoma,Arial,sans-serif;color:black;background-color:white;} b {font-family:Tahoma,Arial,sans-serif;color:white;background-color:#525D76;} p {font-family:Tahoma,Arial,sans-serif;background:white;color:black;font-size:12px;} a {color:black;} a.name {color:black;} .line {height:1px;background-color:#525D76;border:none;}</style></head><body><h1>HTTP Status 404 â?? Not Found</h1><hr class="line" /><p><b>Type</b> Status Report</p><p><b>Message</b> Not found</p><p><b>Description</b> The origin server did not find a current representation for the target resource or is not willing to disclose that one exists.</p><hr class="line" /><h3>Apache Tomcat/8.5.47</h3></body></html>
        at io.supertokens.cli.httpRequest.HTTPRequest.sendGETRequest(HTTPRequest.java:100)
        at io.supertokens.cli.processes.Processes$RunningProcess.fetchConfigFilePath(Processes.java:48)
        at io.supertokens.cli.commandHandler.list.ListHandler.doCommand(ListHandler.java:35)
        at io.supertokens.cli.commandHandler.CommandHandler.handleCommand(CommandHandler.java:31)
        at io.supertokens.cli.Main.start(Main.java:101)
        at io.supertokens.cli.Main.main(Main.java:49)

Could not execute list command.
exiting
```

## Useful informations

Tested on Windows 11

How to reproduce:
1. Open an Administrator command prompt
2. Run `supertokens list` -> `No SuperTokens instances running`
3. Ensure `base_path` in config is commented out
4. Run `supertokens start`
5. Run `supertokens list` -> (proper output)
6. Run `supertokens stop`
7. Modify `base_path` in config to `/your/base/path/here`
8. Run `supertokens start`
9. Run `supertokens list` -> (404 error)

### PR DESCRIPTION
**Pull request by JeremyEastham**
`supertokens list` previously failed when `base_path` was set in `config.yaml` due to a GET request by the CLI that did not observe `base_path`. I added the base path of the server to the .started file in addition to the PID so that the CLI can GET the /config route without encountering 404 errors.

## Summary of change
- Modifies .started file to include the base_path of the server in addition to the PID
- Ensures that the CLI reads this value in order to correctly query the server to get the config file location
- Changes config validation to gracefully handle leading and trailing '/'
- Updates CHANGELOG.md
- Adds relevant tests

## Related issues
- Closes #408
- Replaces #409 

## Test Plan
- Modified `DotStartedFileTest.java` to test the new .started file structure
- Modified `WebserverTest.java` to remove the now irrelevant `base_path` tests

## Documentation changes
(none)

## Checklist for important updates
- [X] Changelog has been updated
    - [X] If there are any db schema changes, mention those changes clearly
- [X] `coreDriverInterfaceSupported.json` file has been updated (if needed)
- [X] `pluginInterfaceSupported.json` file has been updated (if needed)
- [X] Changes to the version if needed
   - In `build.gradle`
- [ ] Had installed and ran the pre-commit hook
- [X] If there are new dependencies that have been added in `build.gradle`, please make sure to add them in `implementationDependencies.json`.
- [X] Issue this PR against the latest non released version branch.
   - To know which one it is, run find the latest released tag (`git tag`) in the format `vX.Y.Z`, and then find the latest branch (`git branch --all`) whose `X.Y` is greater than the latest released tag.
   - If no such branch exists, then create one from the latest released branch.

## Remaining TODOs for this PR
- [X] Ensure tests pass
- [ ] Make sure nothing else writes to/reads from the .started file that might break

### CODE REVIEW

**Comment by JKEKBE** [comment]
@rishabhpoddar I inadvertently closed #409 by force-pulling 3.12 from upstream. I have implemented all of the changes that you suggested and ensured that all tests pass. I decided to remove the '/' checks from the config validation method in favor of gracefully handling these cases in the `getBasePath` method. I also removed the tests that tested the now removed behavior.

File: src/test/java/io/supertokens/test/WebserverTest.java [hunk 1]
```diff
@@ -787,26 +787,6 @@ protected void doGet(HttpServletRequest req, HttpServletResponse resp) throws IO
...
-        {
-            Utils.setValueInConfig("base_path", "/somepath/");
-            String[] args = { "../" };
-            TestingProcess process = TestingProcessManager.start(args);
-            EventAndException e = process.checkOrWaitForEvent(PROCESS_STATE.INIT_FAILURE);
-            assertTrue(e != null && e.exception instanceof QuitProgramException
-                    && e.exception.getMessage().equals("base_path cannot end with '/'"));
-            Utils.reset();
-        }
-
```
Comment (rishabhpoddar) [inline src/test/java/io/supertokens/test/WebserverTest.java:809]: Since these tests are removed, we should add tests which sets the base path as `somepath/` and when calls `Config.getBasePath()`, it returns `"/somepath"`. Likewise for when the basepath is `/somepath/`

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: src/test/java/io/supertokens/test/WebserverTest.java (hunk 1 above)
Comment (JeremyEastham) [reply src/test/java/io/supertokens/test/WebserverTest.java]: I have reimplemented this functionality with calls to `getBasePath` instead of checking the exception.

File: src/test/java/io/supertokens/test/WebserverTest.java [hunk 2]
```diff
@@ -787,6 +788,38 @@ protected void doGet(HttpServletRequest req, HttpServletResponse resp) throws IO
 
     @Test
     public void invalidBasePathTest() throws InterruptedException, IOException {
+        {
+            Utils.setValueInConfig("base_path", "somepath/");
+            String[] args = { "../" };
+            TestingProcess process = TestingProcessManager.start(args);
+            EventAndException e = process.checkOrWaitForEvent(PROCESS_STATE.STARTED);
+            assertEquals(Config.getConfig(process.main).getBasePath(), "/somepath");
+            Utils.reset();
```
Comment (rishabhpoddar) [inline src/test/java/io/supertokens/test/WebserverTest.java:797]:
```suggestion
            process.kill();
            assertNotNull(process.checkOrWaitForEvent(PROCESS_STATE.STOPPED));
            Utils.reset();
```

File: src/test/java/io/supertokens/test/WebserverTest.java [hunk 3]
```diff
@@ -787,6 +788,38 @@ protected void doGet(HttpServletRequest req, HttpServletResponse resp) throws IO
...
+        }
+        {
+            Utils.setValueInConfig("base_path", "/somepath/");
+            String[] args = { "../" };
+            TestingProcess process = TestingProcessManager.start(args);
+            EventAndException e = process.checkOrWaitForEvent(PROCESS_STATE.STARTED);
+            assertEquals(Config.getConfig(process.main).getBasePath(), "/somepath");
+            Utils.reset();
```
Comment (rishabhpoddar) [inline src/test/java/io/supertokens/test/WebserverTest.java:805]:
```suggestion
            process.kill();
            assertNotNull(process.checkOrWaitForEvent(PROCESS_STATE.STOPPED));
            Utils.reset();
```

File: src/test/java/io/supertokens/test/WebserverTest.java [hunk 4]
```diff
@@ -787,6 +788,38 @@ protected void doGet(HttpServletRequest req, HttpServletResponse resp) throws IO
...
+        }
+        {
+            Utils.setValueInConfig("base_path", "somepath");
+            String[] args = { "../" };
+            TestingProcess process = TestingProcessManager.start(args);
+            EventAndException e = process.checkOrWaitForEvent(PROCESS_STATE.STARTED);
+            assertEquals(Config.getConfig(process.main).getBasePath(), "/somepath");
+            Utils.reset();
```
Comment (rishabhpoddar) [inline src/test/java/io/supertokens/test/WebserverTest.java:813]:
```suggestion
            process.kill();
            assertNotNull(process.checkOrWaitForEvent(PROCESS_STATE.STOPPED));
            Utils.reset();
```

File: src/test/java/io/supertokens/test/WebserverTest.java [hunk 5]
```diff
@@ -787,6 +788,38 @@ protected void doGet(HttpServletRequest req, HttpServletResponse resp) throws IO
...
+        }
+        {
+            Utils.setValueInConfig("base_path", "/somepath");
+            String[] args = { "../" };
+            TestingProcess process = TestingProcessManager.start(args);
+            EventAndException e = process.checkOrWaitForEvent(PROCESS_STATE.STARTED);
+            assertEquals(Config.getConfig(process.main).getBasePath(), "/somepath");
+            Utils.reset();
```
Comment (rishabhpoddar) [inline src/test/java/io/supertokens/test/WebserverTest.java:821]:
```suggestion
            process.kill();
            assertNotNull(process.checkOrWaitForEvent(PROCESS_STATE.STOPPED));
            Utils.reset();
```

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

**Comment by rishabhpoddar** [comment]
There are also edge cases where someone can set the `base_path` to `//somepath//`. In which case, things will start to fail.

File: src/test/java/io/supertokens/test/WebserverTest.java [hunk 6]
```diff
@@ -791,30 +792,42 @@ public void invalidBasePathTest() throws InterruptedException, IOException {
...
-
         {
             Utils.setValueInConfig("base_path", "/some path");
             String[] args = { "../" };
             TestingProcess process = TestingProcessManager.start(args);
             EventAndException e = process.checkOrWaitForEvent(PROCESS_STATE.INIT_FAILURE);
             assertTrue(e != null && e.exception instanceof QuitProgramException
                     && e.exception.getMessage().equals("Invalid characters in base_path config"));
-            Utils.reset();
+            process.kill();
```
Comment (rishabhpoddar) [inline src/test/java/io/supertokens/test/WebserverTest.java:830]: - In this case, we expect the core to not start, so don't need to call kill.
- It's a good idea to call ` Utils.reset();` in each of the cases above as that resets the config file to its original form.

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: src/test/java/io/supertokens/test/WebserverTest.java (hunk 6 above)
Comment (JeremyEastham) [reply src/test/java/io/supertokens/test/WebserverTest.java:830]: Since `Utils.reset();` is included in the `beforeEach` method, shouldn't we just delete this line so it's not called twice?

Comment (rishabhpoddar) [reply src/test/java/io/supertokens/test/WebserverTest.java:830]: Yea. We could. But calling it twice is not an issue. I was there originally only so that it matched the "format" of the other test cases in this one test case. But we can remove it.. yea

**Comment by JeremyEastham** [comment]
I'm going to add a test or two for the '//' corner case, but otherwise, I think this PR is good to merge.

**Comment by rishabhpoddar** [comment]
Yea. Those tests, and the other tests you added need to wait until the process is killed before starting the process again.

**Comment by JeremyEastham** [comment]
The code for this method was getting really messy, so I reimplemented it with a HashMap instead of just copying and pasting the same test over and over.

**Comment by JeremyEastham** [comment]
@rishabhpoddar All suggested changes have been implemented and all tests pass now.

**Review by rishabhpoddar — APPROVED** [review]

**Comment by rishabhpoddar** [comment]
thanks @JeremyEastham . Will release this shortly. Also have added your profile to the supertokens-core contributor's list in the readme :)
